File: common/clock/clockflag/time.go
// Copyright 2014 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package clockflag

import (
	"encoding/json"
	"flag"
	"time"
)

// Time is a flag- and JSON-compatible Time which parses from RFC3339 strings.
type Time time.Time

var _ interface {
	flag.Value
	json.Unmarshaler
	json.Marshaler
} = (*Time)(nil)

// Time returns the Time value associated with this Time.
func (t Time) Time() time.Time {
	return time.Time(t)
}

// Set implements flag.Value.
func (t *Time) Set(value string) error {
	timeValue, err := time.Parse(time.RFC3339Nano, value)
	if err != nil {
		return err
	}
	*t = Time(timeValue.UTC())
	return nil
}

func (t *Time) String() string {
	return time.Time(*t).String()
}

func (t Time) Equal(o Time) bool {
	return time.Time(t).Equal(time.Time(o))
}

// UnmarshalJSON implements json.Unmarshaler.
//
// Unmarshals a JSON entry into the underlying type. The entry is expected to
// contain a string corresponding to one of the enum's keys.
func (t *Time) UnmarshalJSON(data []byte) error {
	var value time.Time
	if err := value.UnmarshalJSON(data); err != nil {
		return err
	}
	*t = Time(value.UTC())
	return nil
}

// MarshalJSON implements json.Marshaler.
//
// Marshals a Time into an RFC3339 time string.
func (t Time) MarshalJSON() ([]byte, error) {
	return t.Time().UTC().MarshalJSON()
}

File: logdog/appengine/cmd/coordinator/static/static/js/time.js
// Copyright 2018 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// A Series of time based utilites for the LogDog plaintext viewer.
// Requires: moment.js, moment-timezone.js, jquery, jquery-ui

// This file cannot use EcmaScript6: crbug.com/791528

(function(window) {
  'use strict';

  const tz = moment.tz.guess();

  /**
   * Given a Date, return a time string in the user's local timezone.
   * Also return the time string in relative time from now, MTV time, and UTC
   * time.
   */
  function formatDate(t) {
    var mt = moment.tz(t, tz);
    if (!mt.isValid()) {
      return null;
    }

    return {
      localLong: mt.format("YYYY-MM-DD LTS (z)"),
      MTVLong: moment.tz(mt, "America/Los_Angeles").format(
        "YYYY-MM-DD LTS [(MTV)]"),
      UTCLong: moment.tz(mt, "UTC").format("YYYY-MM-DD LTS [(UTC)]"),
    }
  }

  /***
   * Add local time information as a tooltip.
   */
  function formatTime(node) {
    if (node.title != "") return;  // Already done.
    try {
      var timestamp = node.getAttribute('data-timestamp');
      // Delta since last log line, in milliseconds.
      var deltaS = node.getAttribute('data-delta');
      var delta = parseInt(deltaS, 10) / 1000.0;
      var date = new Date(parseInt(timestamp, 10));
      var newTimestamp = formatDate(date);
      if (newTimestamp != null) {
        node.setAttribute(
          "title", [
            delta.toFixed(2) + "s since last line",
            newTimestamp.localLong,
            newTimestamp.MTVLong,
            newTimestamp.UTCLong,
          ].join("\n")
        )
      }
    } catch (e) {
      console.error('could not convert time of node', node, 'to local:', e)
    }
  }

  function setLocale() {
    // Moment.js does not set the locale automatically, it must be done by the
    // caller.
    var locale = window.navigator.userLanguage || window.navigator.language;
    moment.locale(locale);
  }

  // Export all methods and attributes as module level functions.
  Object.assign(window.utils = window.utils|| {}, {
    formatTime: formatTime,
    setLocale: setLocale,
  });

}(window));
